src/java.base/share/classes/jdk/internal/foreign/CaptureStateUtil.java

```diff
@@ -0,0 +1,396 @@
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.foreign;
+
+import jdk.internal.invoke.MhUtil;
+import jdk.internal.lang.stable.InternalStableFactories;
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.Linker;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.StructLayout;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.VarHandle;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+
+/**
+ * An internal utility class that can be used to adapt system-call-styled method handles
+ * for efficient and easy use.
+ */
+public final class CaptureStateUtil {
+
+    private static final StructLayout CAPTURE_LAYOUT = Linker.Option.captureStateLayout();
+    private static final BufferStack POOL = BufferStack.of(CAPTURE_LAYOUT);
+
+    // Do not use a lambda in order to allow early use in the init sequence
+    private static final Function<BasicKey, MethodHandle> UNDERLYING_MAKE_BASIC_HANDLE = new Function<>() {
+        @Override
+        public MethodHandle apply(BasicKey basicKey) {
+            return makeBasicHandle(basicKey);
+        }
+    };
+
+    // The `BASIC_HANDLE_CACHE` contains the common "basic handles" that can be reused for
+    // all adapted method handles. Keeping as much as possible reusable reduces the number
+    // of combinators needed to form an adapted method handle.
+    // The function is lazily computed.
+    //
+    private static final Function<BasicKey, MethodHandle> BASIC_HANDLE_CACHE;
+
+    static {
+        final Set<BasicKey> inputs = new HashSet<>();
+        // The Cartesian product : (int.class, long.class) x ("errno", ...)
+        // Do not use Streams in order to enable "early" use in the init sequence.
+        for (Class<?> c : new Class<?>[]{int.class, long.class}) {
+            for (MemoryLayout layout : CAPTURE_LAYOUT.memberLayouts()) {
+                inputs.add(new BasicKey(c, layout.name().orElseThrow()));
+            }
+        }
```

Comment:
```suggestion
        for (MemoryLayout layout : CAPTURE_LAYOUT.memberLayouts()) {
            String name = layout.name().orElseThrow();
            inputs.add(new BasicKey(int.class, name));
            inputs.add(new BasicKey(long.class, name));
        }
```
There are only two classes, int.class and long.class, so this might be simpler.